lib/streams.gi

```diff
@@ -244,6 +243,104 @@ InstallOtherMethod( ReadAsFunction,
     [ IsInputStream ],
     READ_AS_FUNC_STREAM );
 
+#############################################################################
+##
+#F  ReadToList()  . . . . . . .  Reads commands and returns a list of results
+##
+##  <#GAPDoc Label="ReadToList">
+##  <ManSection>
+##  <Func Name="ReadToList" Arg='stream[,output_stream]'/>
+##
+##  <Description>
+##    Reads <A>stream</A> which is either an InputStream or a string containing
+##    complete GAP commands, executes them, and returns a list of results.
+##    The list of results is organized like the result of <C>READ_ALL_COMMANDS</C>.
+##    
+##    If the optional output stream <A>output_stream</A> argument is given, all
+##    additional output that happens while executing those commands is redirected
+##    into <A>output_stream</A>. This includes both standard and error output.
+##    If <A>output_stream</A> is <C>true</C>, the additional output is printed.
+##  </Description>
+##  </ManSection>
+##  <#/GAPDoc>
+##
+BindGlobal("ReadToList",
+function(stream, output_stream...)
+    local opened_stream, return_list, opened_dummy_output;
+
+    if IsString(stream) then
+        stream := InputTextString(stream);
+        opened_stream := true;
+    else
+        opened_stream := false;
+    fi;
+
+    opened_dummy_output := false;
+    if Length(output_stream) = 0 or output_stream[1] = false then
+        output_stream := OutputTextNone();
+        opened_dummy_output := true;
+    elif IsOutputStream(output_stream[1]) then
+        output_stream := output_stream[1];
+    elif output_stream[1] = true then
+        output_stream := 0;
+    else
+        Error("second argument must be an output stream or a boolean");
+    fi;
+
+    return_list := READ_ALL_COMMANDS(stream, output_stream, false);
+
+    if opened_stream then
+        CloseStream(stream);
+    fi;
+    if opened_dummy_output then
+        CloseStream(output_stream);
+    fi;
+
+    return return_list;
+end );
+
+#############################################################################
+##
+#F  ReadToString()  . . . . . . . . . . . Reads commands and returns a string
+##
+##  <#GAPDoc Label="ReadToString">
+##  <ManSection>
+##  <Func Name="ReadToString" Arg='stream[,output_stream]'/>
+##
+##  <Description>
+##    The same as <Ref Func="ReadToList"/>, except that it executes
+##    <Ref Oper="ViewObj"/> on every result and returns the resulting
+##    string.
```

Comment:
If I hadn't been in a discussion on this last week, I don't think I'd understand this. For starters, in which order are the commands executed and ViewObj called? Interleaved, as the REPL would do? Or are first all commands executed, then all results printed? Looking at the code, it seems like the latter; note that this can lead to different results compared to the former (e.g. if printing the objects causes additional computations, such as computing the size, which then affects subsequent commands).

Also, what is "the resulting string? After all, `ViewObj` returns nothing. So perhaps this:

> This invokes ReadToList, and then calls ViewObj on every result, but with the output redirected to a string, which is then returned.